### PR TITLE
fix: clean up data handling in TFKerasTrial [DET-4300]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Stop using the TF OrderedEnqueuer; write our own that is far simpler and
supports reproducible shuffling. Implementing reproducible shuffling
within the Sequence inside of the real OrderedEnqueuer is technically
possible but difficult and fragile.

Rely on new callback support for num_inputs.

Delete the now-unneeded input manager objects.

## Commentary

This is a totally different approach than https://github.com/determined-ai/determined/pull/1545.  Compared to that change, this change has the following pros/cons:

**Pros** of this approach vs #1545:

- does not require intercepting calls to reset_metrics
- does not allow the model.fit() loop to cycle through epochs (we handle epoch ends ourselves, so if model.fit is also trying to handle them we have to neutralize all side-effects of those operations)
- does not include the hard-to-understand `_DeterminedSequenceWrapper`, which tried to do shuffling in the Sequence without support from the OrderedEnqueuer
- the `_Sampler` part of the `_Enqueuer` has approximately the same semantics as the pytorch Sampler, which makes repeatable shuffles way easier.

**Cons** of this approach vs #1545:

- includes a custom replacement of the Keras OrderedEnqueuer.  Ours is many hundreds of lines shorter than the original one, and uses python standard libraries at least, but nonetheless it is more code.

I would actually argue this version lowers our surface area for support, since the repeatable shuffle built on the real OrderedEnqueuer in #1545 depends on several implementation details of the original OrderedEnqueuer in order to function, and had several workarounds for limitations in the original OrderedEnqueuer (like tracking two epochs of indices at all time).  That made our feature fragile, considering how unstable TF code has proven to be over time.  This version interacts with more via public APIs and with less reliance on keras internals.

### CODE REVIEW

**Comment by rb-determined-ai** [comment]
Just finished running the cifar 10 example with this version and observed absolutely no slowdown after 32 epochs (I observed a linear slowdown on tip of master and a very slight slowdown with #1545)

File: harness/determined/keras/_enqueuer.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        sequence: tf.keras.utils.Sequence,
+        sampler: _Sampler,
+        repeat: bool,
+        workers: int,
+        depth: int,
+    ):
+        self.sequence = sequence
+        self.sampler = sampler
+        self.repeat = repeat
+        self.depth = depth
```
Comment (aaron276h) [inline harness/determined/keras/_enqueuer.py:129]: non-blocking: Probably should add a check to make sure `depth` > 0

File: harness/determined/keras/_tf_keras_trial.py [hunk 2]
```diff
@@ -596,29 +589,78 @@ def run(self) -> None:
...
+            self._stop_enqueuers()
 
     def _launch_fit(self) -> None:
-        (
-            training_input,
-            batches_per_epoch,
-        ) = self._train_input_manager.get_training_input_and_batches_per_epoch()
+        training_data = self.training_data
+
+        if isinstance(training_data, keras.SequenceAdapter):
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_trial.py:598]: non-blocking: If this is only for backwards compatibility can we leave a comment saying so

**Review by aaron276h — COMMENTED** [review]
Looks really good. I do think it's worthwhile spending some time adding comments (and a few variable renamings) to make it easier to maintain `_enqueuer.py` going forward.

**Comment by shiyuann** [comment]
@shiyuann I have reviewed the code except for the test code. All the other things except for the implementation of enqueuer and its inherited classes look good. I am confused about the high-level communication pattern. Specific questions include: how do enqueuers communicate with the training thread/process? What do the `query` and `answer` mean?

**Review by rb-determined-ai — COMMENTED** [review]
I'm closing the other one and I'm going to move forward with this one.

File: harness/determined/keras/_enqueuer.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    def __init__(
+        self,
+        length: int,
+        shard_rank: int,
+        shard_size: int,
+        shuffle: bool,
+        shuffle_seed: int,
+        prior_batches_trained: int,
+    ) -> None:
+        self.indices = list(range(length))
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply harness/determined/keras/_enqueuer.py:38]: Thanks for adding that

File: harness/determined/keras/_tf_keras_trial.py [hunk 4]
```diff
@@ -384,6 +367,62 @@ def __init__(
...
+        # Non-tf.data.Datasets should not have used the data layer.
+        if not isinstance(self.training_data, tf.data.Dataset):
+            if cacheable_used:
+                raise det.errors.InvalidExperimentException(
+                    "Pass in a tf.data.Dataset object for training data if using "
+                    "context.experimental.cache_train_dataset().",
+                )
+            return
+
+        # You can't use data layer and the wrap_dataset.w
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_trial.py:383]: typo

File: harness/determined/keras/_enqueuer.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        self.queries = self.queue_class()()
+        self.answers = self.queue_class()()
+
+        self.workers = [
+            self.worker_class()(target=_worker, args=(self.sequence, self.queries, self.answers))
+            for _ in range(workers)
+        ]
+
+    def start(self) -> None:
+        assert not self.started and not self.stopped, "restarting an enqueuer is not allowed"
```
Comment (aaron276h) [inline harness/determined/keras/_enqueuer.py:229]: non-blocking: this should probably be a `check`

File: harness/determined/keras/_enqueuer.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+
+def _worker(sequence: tf.keras.utils.Sequence, queries: Queue, answers: Queue) -> None:
+    """
+    _worker defines a data loader worker's primary loop.  This loop runs in either a
+    threading.Thread or a multiprocessing.Process; the caller (a _WorkerEnqueuer) is responsible
+    for passing in the appropriate type of Queue.
+
+    Parameters:
+        sequence: the user-provided Keras Sequence
```
Comment (aaron276h) [inline harness/determined/keras/_enqueuer.py:158]: missing some periods 😱

File: harness/determined/keras/_tf_keras_context.py [hunk 7]
```diff
@@ -56,6 +56,7 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
         self._fit_workers = 1
         self._fit_use_multiprocessing = False
         self._fit_max_queue_size = 10
+        self._fit_shuffle = True
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:59]: praise: thank you for matching the tf.keras fit()0 default

**Review by aaron276h — APPROVED** [review]
Really nice work!

File: harness/determined/keras/_enqueuer.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_enqueuer.py:229]: I disagree, as it is an internal object and the assertion is validating code structure, not user inputs, and the check module would not print anything useful anyway.

File: harness/determined/keras/_tf_keras_context.py (hunk 7 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:59]: Yeah it's about time we supported shuffle haha

File: harness/determined/keras/_tf_keras_trial.py [hunk 8]
```diff
@@ -609,30 +589,81 @@ def run(self) -> None:
...
+            training_data,
             class_weight=self.context._fit_class_weight,
             callbacks=self.callback_list,
             shuffle=False,
             steps_per_epoch=sys.maxsize,
-            initial_epoch=self._train_input_manager.get_initial_epoch(),
             epochs=IMPOSSIBLY_LARGE_EPOCHS,
             validation_split=0,
             verbose=0,
+            workers=0,
```
[the remark this answers is not in the document]
Comment (shiyuann) [reply harness/determined/keras/_tf_keras_trial.py:703]: non-blocking: It would be good to add a comment here.

File: harness/determined/keras/_enqueuer.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        # Example:
+        #   let length=10, shard_rank=0, and shard_size=3:
+        #   epoch 1: 0, 3, 6, 9
+        #   epoch 2: 2, 5, 8
+        #   epoch 3: 1, 4, 7
+        #   epoch 4: (same as epoch 1)
+        # In this example, the offset in the first three epochs is 0, then 2, then 1.
+        self.offset = shard_rank
+
+        if self.shuffle:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_enqueuer.py:61]: > What do you mean? We can't save the state but it would be easy to calculate the state?

Due to the prefetching, there are two timelines going on.  The Sampler might think it is a couple batches into epoch 9 while the model thinks it is a couple of batches from the end of epoch 8.  You always decide to save the state of the model on the model's timeline, but there's no guarantee that if you save the state of the RNG in the Sampler right then that you are getting the correct RNG state.  This is the same reason why we need to implement a distributed snapshot for master restart; because the searcher and the workload sequencers are each on different timelines.

However, it is easy to just start the Sampler from epoch 0 every time and call `_end_epoch()` to the exact number of epochs that the Sampler would have gone through, based on the prior batches processed for the initial workload.  At that point, the next index from the Sampler will be the next index that the model needs.  That's what I mean when I say it's easy to calculate the state.  This is identical to what we do in `SkipBatchSampler` in the pytorch code.

File: harness/determined/keras/_enqueuer.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+            self._end_epoch()
+
+        self.offset += self.num_shards * batches_to_skip
+
+    def _this_epoch_indices(self) -> range:
+        return range(self.offset, len(self.indices), self.num_shards)
+
+    def _end_epoch(self) -> None:
+        """
+        The _Sampler is stateful, and _epoch_end is where it modifies it state after each epoch.
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_enqueuer.py:79]: It does not save any state.

File: harness/determined/keras/_enqueuer.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+            if query is None:
+                return
+            i, order = query
+            data = sequence[i]
+            answers.put((data, order))
+    finally:
+        answers.put(None)
+
+
+class _WorkerEnqueuer(_Enqueuer):
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_enqueuer.py:175]: Sure

File: harness/determined/keras/_tf_keras_trial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_trial.py:703]: I added a comment in the place where we handle the args, rather than here (where we only override some args, and leave others as their defaults).

**Review by shiyuann — APPROVED** [review]
good work! sorry, it takes me a long time to review.

**Comment by rb-determined-ai** [comment]
@shiyuann on the contrary, thank you for a careful review!
